documentation/prometheus-design-principles.md

```diff
@@ -0,0 +1,164 @@
+    lives
+
+The
+[prometheus docs also say](https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push?):
+
+> Overall, we believe that pulling is slightly better than pushing,
+> but it should not be considered a major point when considering a
+> monitoring system.
+
+### When to use pushgateway
+
+There's a
+[page in the docs for when to use pushgateway](https://prometheus.io/docs/practices/pushing/).
+
+## Shape of data
+
+### Names
+
+  * tags should not have too high a cardinality
+      * See the CAUTION here: https://prometheus.io/docs/practices/naming/#labels
+      * it seems that prometheus 2.0's new storage engine took a lot
+      of effort to support "churn" of tag labels, so long as the
+      number of active labels at any point is relatively small. see
+      ["storing 16 bytes at scale" from promcon 2017](https://promcon.io/2017-munich/talks/storing-16-bytes-at-scale/)
+      or
+      [prometheus 2.0: new storage layer...](https://coreos.com/blog/prometheus-2.0-storage-layer-optimization)
+  * tags should not include redundant informational data
+
+### Values
+
+  * [counters should always count up from zero](https://www.robustperception.io/how-does-a-prometheus-counter-work/)
+      * data should always make sense, independent of scrape interval
+      * this is in contrast to statsd's behaviour
+  * [sum of rates, not rate of sums](https://www.robustperception.io/rate-then-sum-never-sum-then-rate/)
+
+## Architectural things ##
+
+### Simplicity ###
+
+Prometheus is designed to achieve reliability through simplicity.
+
+This means:
+  * redundancy is achieved by deploying multiple separate instances,
+    not by database replication.
+  * recovery from backup is not a priority.  if a prometheus server
+    has a disk failure, just use another redundant instance (assuming
+    you have one)
+  * if a prometheus instance has a temporary blip (say, a network
+    issue) and then recovers, it will have missing metrics for the
+    duration of the issue.
+    
+[Julius Volz talked about why they didn't do clustering](https://youtu.be/QgJbxCWRZ1s?t=26m40s).
+
+### Prometheus should be deployed in same failure domain ###
+
+A prometheus server should be deployed in the same failure domain as
+the things it monitors.  This means the same data centre, region,
+availability zone, etc.
+
+This means that:
+
+  * a prometheus shouldn't have to go over the public internet to
+    scrape a service
+      * (federation might go over public internet, though)
+      
+https://www.robustperception.io/scaling-and-federating-prometheus/
+
+### There should normally be a single alertmanager for an organization ###
+
+It seems that there should normally be a single alertmanager for an
+organization.  The best reference I've found for this is from
+[Julius Volz's Prometheus Design and Philosophy talk (7:44)](https://youtu.be/QgJbxCWRZ1s?t=7m43s)
+which doesn't go into much detail as to *why* you'd do this.
+
```

Comment:
My guess: by having multiple alertmanager instances one runs the risk of multiple alerts firing for the same symptom. Which goes against one of the main purposes of having alertmanager: deduplication of alerts.
They appear to be working on a cluster for multiple instances to improve the availability of that one instance. We'd need to test the behaviour of this before endorsing it though given its slightly "experimental" nature.

Reply:
I didn't write very clearly here but I think there's two different issues:

 - multiple alertmanagers for redundancy
 - multiple alertmanagers for different configuration for different teams

I think the first is a thing we should consider (subject to the warnings you mention).  I think the second is generally not done.